src/adapter/src/coord/introspection.rs

```diff
@@ -32,16 +33,56 @@ use crate::notice::AdapterNotice;
 use crate::session::Session;
 use crate::{rbac, AdapterError};
 
+/// Returns whether a scalar expression can run an expensive function
+fn scalar_can_run_expensive_function(scalar: &MirScalarExpr) -> bool {
+    match scalar {
+        mz_expr::MirScalarExpr::Column(_)
+        | mz_expr::MirScalarExpr::Literal(_, _)
+        | mz_expr::MirScalarExpr::CallUnmaterializable(_) => false,
+        mz_expr::MirScalarExpr::CallUnary { .. }
+        | mz_expr::MirScalarExpr::CallBinary { .. }
+        | mz_expr::MirScalarExpr::CallVariadic { .. } => true,
+        mz_expr::MirScalarExpr::If { cond, then, els } => {
+            scalar_can_run_expensive_function(cond)
+                || scalar_can_run_expensive_function(then)
+                || scalar_can_run_expensive_function(els)
+        }
+    }
+}
+
+/// Returns whether a expression can run an expensive function. We err on returning false.
+/// The goal is to be a heuristic for "this expression is cheap to evaluate:
+/// ideally not creating a dataflow at all when optimized".
+fn expr_can_run_expensive_function(expr: &MirRelationExpr) -> bool {
+    // FlapMap has a table function while all other constructs use MirScalarExpr to run a function
```

Comment:
yeah that's a good point. added to the impl but since that file is pretty big I don't have a lot of faith someone would audit it. There are the introspection tests that may catch some mistakes